Plugins/NativeServerPlugin/src/directx_buffer_capturer.cpp

```diff
@@ -0,0 +1,258 @@
+
+using namespace Microsoft::WRL;
+using namespace StreamingToolkit;
+
+DirectXBufferCapturer::DirectXBufferCapturer(ID3D11Device* d3d_device) :
+	d3d_device_(d3d_device)
+{
+}
+
+void DirectXBufferCapturer::Initialize(bool headless, int width, int height)
+{
+	// Gets the device context.
+	d3d_device_->GetImmediateContext(&d3d_context_);
+
+#ifdef MULTITHREAD_PROTECTION
+	// Enables multithread protection.
+	ID3D11Multithread* multithread;
+	d3d_device_->QueryInterface(IID_PPV_ARGS(&multithread));
+	multithread->SetMultithreadProtected(true);
+	multithread->Release();
+#endif // MULTITHREAD_PROTECTION
+
+	// Initializes NVIDIA encoder.
+	webrtc::H264EncoderImpl::SetDevice(d3d_device_.Get());
+	webrtc::H264EncoderImpl::SetContext(d3d_context_.Get());
+
+	// Headless mode initialization.
+	headless_ = headless;
+	if (headless_)
+	{
+		ResizeRenderTexture(width, height);
+	}
+}
+
+void DirectXBufferCapturer::SendFrame(int64_t prediction_time_stamp)
+{
+	if (!headless_)
+	{
+		return;
+	}
+
+	SendFrame(render_texture_.Get(), prediction_time_stamp);
+}
+
+void DirectXBufferCapturer::SendFrame(ID3D11Texture2D* frame_buffer, int64_t prediction_time_stamp)
+{
+	if (!running_)
+	{
+		return;
+	}
+
+	// Updates staging frame buffer.
+	UpdateStagingBuffer(frame_buffer);
+
+	// Creates webrtc frame buffer.
+	D3D11_TEXTURE2D_DESC desc;
+	staging_frame_buffer_->GetDesc(&desc);
+	rtc::scoped_refptr<webrtc::I420Buffer> buffer = 
+		webrtc::I420Buffer::Create(desc.Width, desc.Height);
+
+	// For software encoder, converting to supported video format.
+	if (use_software_encoder_)
+	{
+		D3D11_MAPPED_SUBRESOURCE mapped;
+		if (SUCCEEDED(d3d_context_.Get()->Map(
+			staging_frame_buffer_.Get(), 0, D3D11_MAP_READ, 0, &mapped)))
+		{
+			libyuv::ABGRToI420(
+				(uint8_t*)mapped.pData,
+				desc.Width * 4,
+				buffer.get()->MutableDataY(),
+				buffer.get()->StrideY(),
+				buffer.get()->MutableDataU(),
+				buffer.get()->StrideU(),
+				buffer.get()->MutableDataV(),
+				buffer.get()->StrideV(),
+				desc.Width,
+				desc.Height);
+
+			d3d_context_->Unmap(staging_frame_buffer_.Get(), 0);
+		}
+	}
+
+	// Updates time stamp.
+	auto time_stamp = std::chrono::duration_cast<std::chrono::nanoseconds>(
+		std::chrono::system_clock::now().time_since_epoch()).count();
+
+	// Creates video frame buffer.
+	auto frame = webrtc::VideoFrame(buffer, kVideoRotation_0, time_stamp);
+	frame.set_ntp_time_ms(clock_->CurrentNtpInMilliseconds());
+	frame.set_rotation(VideoRotation::kVideoRotation_0);
+	frame.set_prediction_timestamp(prediction_time_stamp);
+
+	// For hardware encoder, setting the video frame texture.
+	if (!use_software_encoder_)
+	{
+		frame.set_staging_frame_buffer(staging_frame_buffer_.Get());
+	}
+
+	// Sending video frame.
+	BufferCapturer::SendFrame(frame);
+}
+
+void DirectXBufferCapturer::SendFrame(ID3D11Texture2D* left_frame_buffer, ID3D11Texture2D* right_frame_buffer, int64_t prediction_time_stamp)
+{
+	if (!running_)
+	{
+		return;
```

Comment:
if this is code that we'd expect another developer to call i'd much prefer to see some error condition occur here instead of an early return...since it seems the caller should not call this signature unless they are running. that way the dev can better diagnose what's going on

Reply:
Same as above.